lib/active_merchant/billing/gateways/payflow.rb

```diff
@@ -188,6 +189,53 @@ def build_credit_card_request(action, money, credit_card, options)
           end
         end
         xml.target!
+        add_level_two_three_fields(xml, options)
+      end
+
+      def add_level_two_three_fields(xml, options)
+        if options[:level_two_fields] || options[:level_three_fields]
+          xml_doc = Nokogiri::XML.parse(xml.target!) ## so it doesn't put on an xml instruction
+          %i[level_two_fields level_three_fields].each do |fields|
+            xml = add_fields(xml_doc, options[fields]) if options[fields]
+          end
+        end
+        xml
+      end
+
+      def check_fields(parent, fields, xml_doc)
+        fields.each do |k, v|
+          if v.is_a? String
+            new_node = Nokogiri::XML::Node.new(k, xml_doc)
+            new_node.add_child(v)
+            xml_doc.at_css(parent).add_child(new_node)
+          else
+            check_subparent_before_continuing(parent, k, xml_doc)
+            check_fields(k, v, xml_doc)
+          end
+        end
+        xml_doc
+      end
+
+      def check_subparent_before_continuing(parent, subparent, xml_doc)
+        unless xml_doc.at_css(subparent)
+          subparent_node = Nokogiri::XML::Node.new(subparent, xml_doc)
+          xml_doc.at_css(parent).add_child(subparent_node)
+        end
+      end
+
+      def string_on_fields(xml_doc, fields)
+        check_fields('Invoice', fields, xml_doc)
+        xml_doc.root.to_s
+      end
+
+      def add_fields(xml_doc, options_fields)
+        if options_fields.is_a? Hash
+          fields = options_fields
+        elsif options_fields.is_a? String
+          fields = JSON.parse(options_fields)
+        end
+        return unless fields
+        string_on_fields(xml_doc, fields)
```

Comment:
Sorry if you get this comment twice, I think something got messed up on my end.

I would recommend simplifying this `add_fields` method if possible. Since `options_fields` is either a Hash or a String, and won't be `nil` since you guard against that before the call from `add_level_two_three_fields`, we can just use `options_fields` directly, parsing it first into a Hash if necessary.

One idea:

```
      def add_fields(xml_doc, options_fields)
        if options_fields.is_a? String
          options_fields = JSON.parse(options_fields)
        end
        string_on_fields(xml_doc, options_fields)
      end
```